keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,404 @@
+- If the block times out, skip the cache and serve the request directly from storage (etcd).
+
+To get the revsion we have some options: 
+
+- Use an etcd range request with `WithCount` enabled so etcd return only a count and revision 
+- Use an etcd range request against a known empty range with limit=1 as an additional guard (since etcd does not allow for limit=0)
+
+We will explore how long the kube-apiserver should wait for a watch cache to
+catch up to the needed revision. I.e. for our chosen progress event interval,
+should the kube-apiserver also wait a similar interval for the required revision
+to become available or should it fallback to making the request to etcd?
+
+Potential optimiation: We could delay requests, accumulate multiple in-flight
+read requests over some short time period, and at the end of the period, get the
+current revision from etcd, wait for the watch cache to catch up, and then serve
+all the the in-flight reads from cache. This would reduce the number of "get
+current revision" requests that need to be made to etcd in exchange for higher
+request latency (but only for consistent reads). A simple implementation would
+be to do this on a fix interval, where, obviously, if there were not reqests
+during the period, we don't bother to fetch a current revision from etcd.
+
+Optional: For some (but not all) of the etcd progress watch events, also create a
+kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability). See [Watch
+Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
+for details.
+
+#### Determining if etcd is sending progress notify events
+
+It is possible to automatically determine if etcd is sending progress notify events.
+Each watch cache could keep track of when it received the last progress notify event.
+If it has been sufficiently long since the last one was received, or if none have
+ever been received, the watch cache should assume etcd is not sending progress notify,
+and not attempt to serve consistent reads from cache, falling back to serving
+the reads directly from etcd.
+
+### Risks and Mitigations
+
+Configuring etcd to send progress notify events may have performance implications.
+@mm4tt and @wojtek-t have run [some
+experiments](https://github.com/kubernetes/kubernetes/pull/86769) that suggest
+that a progress notify interval of 50ms results in a noticeable increase in both
+etcd and kube-apiserver CPU utilization and some increased kube-apiserver serving
+latency. We intend to address this by first trying a 250ms progress notify interval,
+if ths performs well we will use that as our interval. If not, we will dig in
```

Comment:
> don't buy @lavalamp `s arguments that with Priority and Fairness we will have many requests waiting in queues for long times. In some cases for very overloaded clusters - sure

The argument is basically:
* Right now, everyone leaves a ton of extra resources for the control plane, because if you exceed your limits, the cluster more or less becomes unusable.
* Once that is no longer true, people will size their control plane to match their usage (for cost reasons if no other)
* Maximum efficiency implies non-zero queue lengths
* Many of our requests are < 100ms
* Queue length can be ~10 without breaking our SLO very often
* Non-important requests don't need anything like our SLO, for those, the queue length can be much longer

Basically, it will make all clusters into "heavily loaded" ones, since it will make heavily loaded clusters usable.

Reply:
Added a short section about the reasoning for going with 250ms.

Reply:
@lavalamp - I don't agree with the argument. Mostly with this one:

> Once that is no longer true, people will size their control plane to match their usage (for cost reasons if no other)

From scalability POV, I will never say that people should now allocated less resources for control plane. What the priority & fairness will give us imho is not the ability to save a bit on control plane resources. but rather the fact that the cluster won't completely blow up when either (a) some spike in load or (b) something bad causing cascading failure will happen in the cluster.

I can imagine that some people will decrease amount of resources for control plane for small clusters (where those are significant percentage). But for the largest ones (which is where this will help most and those where we can potentially have biggest issues), I bet it won't happen (as the gain from it is also close to negligible).

Reply:
> But for the largest ones (which is where this will help most and those where we can potentially have biggest issues), I bet it won't happen (as the gain from it is also close to negligible).

I guess I am one of the people who run the largest clusters. We run control plane on beefy machines. Even with priority/fairness support, more efficient caching etc., it will not change. The resource overhead for a large cluster is negligible. Over provisioning gives us the flexibility and time to react for unexpected problems.

Reply:
Yeah - that's exactly what I had on my mind writing my above comment. Thanks @xiang90